VSR: Retain fault information during on_start_view

sentientwaffle opened the pull request:
Found by VOPR seed `12162487559395126725` on commit `5cebb0d9488e8da89681fad084d43035c697dc05`.
Manifested as a panic at:

    // Never nack op_head_min (aka commit_max).
    assert(op > op_head_min);

But I think it could be a correctness bug depending on the state of the other replicas that participate in the view-change.

## Pre-merge checklist

Performance:

* [x] I am very sure this PR could not affect performance.

sentientwaffle committed:
VSR: Retain fault information during on_start_view

Found by VOPR seed `12162487559395126725` on commit `5cebb0d9488e8da89681fad084d43035c697dc05`.
Manifested as a panic at:

    // Never nack op_head_min (aka commit_max).
    assert(op > op_head_min);

But I think it could be a correctness bug depending on the state of the other replicas that participate in the view-change.

matklad commented on src/vsr/replica.zig:
```diff
@@ -1665,7 +1665,25 @@ pub fn ReplicaType(
+                    // 6. Replica A participates in a view-change.
+                    // During this view-change, replica A *must not* nack op X.
+                    // Otherwise, op X might be truncated from the new view.
+                    // To ensure this, leave the entry as faulty.
```
Could the same scenario happen in `primary_set_log_from_do_view_change_messages`?

If `4` is

* Replica A recovers into `.view_change`
* Replica `A` receives DVC quorum and bemoves primary
* Replica replaces `X` via canonical headers
* Another view change happens, and replica `A` erroneously nacks `X`

matklad replied:
Hm, actually, shouldn't

```
            if (journal.has(header)) {
                assert(journal.dirty.bit(slot));
                // Do not clear any faulty bit for the same entry.
            } else {
```

logic in Journal already handle both of these cases?

sentientwaffle replied:
To answer your second question — that journal logic doesn't always help.
In the example I put in the code, we have no header at all (`command=reserved`) prior to calling `replace_header` (the WAL was corrupt).

For your former question:
We could put this same logic in `primary_set_log_from_do_view_change_messages`, but I don't _think_ it is necessary.
The difference is that the headers of this new view cannot be lost, even if we crash + recover the WAL — because they are stored in our superblock's `vsr_headers`, and restored (if necessary) during `open`.

matklad replied:
:thinking: tbh, I am not a fan of this save-restore logic, especially given that it papers over similar logic on journal. 

Could we do something like

```zig
            if (journal.has(header) or journal.faulty(slot) and journal.headers[slot].command == .reserved) {
                assert(journal.dirty.bit(slot));
                // Do not clear any faulty bit for potentially the same entry.
```

in the journal perhaps?

sentientwaffle replied:
I need to play with this a bit, I think there are some tricky cases.
I think we need to leave `set_header_as_dirty` as-is since leaving faulty intact in all cases could impact availability.
But we could add a new function that does leave faulty intact, which the replica can use in some circumstances. (Hand-waiving here b/c I'm not sure which all circumstances those are yet...)

Also I think we can't check `command == reserved` because `Replica.open` loads headers from vsr_headers but leaves `faulty` intact.

sentientwaffle committed:
VSR/Test: Add regression test for corrupt-pipeline-nack

sentientwaffle committed:
VSR: set_header_as_dirty retain faulty; conservative open

Two-part fix to replace the kludge/fix in `on_start_view()`.

A. `Journal.set_header_as_dirty()` now retains the faulty bit *iff* the header was previously reserved.
B. `Replica.open()` loads *at most one* header from `vsr_headers` into its journal.

*A* means we don't unset faulty when the slot's owner might not have actually changed.
That is, when we `set_header_as_dirty(H)`, if `H` might have been in the slot before being corrupted, don't unset faulty.
(This ensures that we do not nack H incrrectly.)
Only unset faulty when we are certain the header changed.

*B* is necessary to avoid the following scenario:
1. Replica A prepares[/commits] op X.
2. Replica A crashes.
3. Prepare X is corrupted in the WAL.
4. Replica A recovers. During `Replica.open()`, Replica A loads the header for op `X - journal_slot_count` (same slot, prior wrap) from vsr_headers into the journal.
5. Replica A participates in a view-change, but nacks[/does not include] op X.
6. Op X is truncated.

sentientwaffle commented:
Updated with a different approach.
This is a more complicated change but the end result is a little simpler imo.

See commit message for explanation.

matklad commented on src/vsr/replica.zig:
```diff
@@ -440,77 +440,69 @@ pub fn ReplicaType(
 
-            var op_head = for (vsr_headers.slice) |*header| {
-                if (vsr.Headers.dvc_header_type(header) != .valid) continue;
-                if (header.op <= self.op_checkpoint_trigger()) break header.op;
-            } else op_head: {
+                    if (!self.journal.has(vsr_header)) {
+                        self.journal.set_header_as_dirty(vsr_header);
+                    }
+                    break;
```
Maybe 

```
                    if (!self.journal.has(vsr_header)) {
                        assert(op_head == null or op_head.? < vsr_header.op)
                        self.journal.set_header_as_dirty(vsr_header);
                    }
                   op_head = vsr_header.op;

```

?

To double-check that vsr header and journal header don't disagree

sentientwaffle replied:
:thinking: I think that assertion is actually not right though — if `op_head.? == vsr_header.op`, then we do want to `set_header_as_dirty` to replace the existing journal header with the one from vsr_headers. That is, our WAL has an old version of the op, but we serialized a view change with a newer header, but did not yet repair the WAL, then crashed.

matklad replied:
Ah, right, journal and  SV may actually diagree with each other!

matklad commented on src/vsr/replica.zig:
```diff
@@ -440,77 +440,69 @@ pub fn ReplicaType(
+            }
+
+            // Try to use vsr_headers to update our head op.
+            // Loading headers prior to the head would risk the following scenario:
```
This comment is a bit confusing (at least, I didn't get it until I also read the commit message). Maybe 

```
            // Try to use vsr_headers to update our head op and its header.
            // Don't load headers prior to the head to avoid the following scenario:
```

matklad reviewed (dismissed):
Two nits, otherwise LGTM!

sentientwaffle committed:
VSR: Clarify comment
